Admin select component performance

chaimann opened the pull request:
## Summary

Allows to fetch options from remote resource, rather than rendering all options in place. Additionally, if the endpoint provides pagination, loads options page by page (using "Link" response header `rel="next"`)

### Usage
#### `:src` attribute
When initialising select component pass `:src` attribute with a url to JSON resource, for `:choices` pass the values you want to be preselected or an empty array.
#### Mapping fields
Maps fields "id" and "name" from JSON data to be value and text of an option accordingly. If the fields have different names, map them with `:"data-option-value-field"` and `:"data-option-label-field"` when initialising select component, e.g.:
```js
// JSON response
[
  { "key": "ruby", "text": "Ruby" },
  { "key": "js", "text": "JavaScript" }
]
```
```ruby
# to render these as options use:
component("ui/forms/select").new(
  ...,
  "data-option-value-field": "key",
  "data-option-label-field": "text",
)
```
#### Nested JSON data
If data in response JSON is nested, you can specify `:"data-json-path"` parameter to retrieve the data by given path signature, e.g.:
```js
// JSON response
{
    "request": "https:\/\/whatcms.org\/API\/List",
    "result": {
        "code": 200,
        "msg": "Success",
        "list": [
            {
                "id": 1,
                "label": "WordPress",
                "type": [
                    "Blog",
                    "CMS"
                ]
            },
            {
                "id": 2,
                "label": "Joomla",
                "type": [
                    "Other CMS",
                    "CMS"
                ]
            },
    ...
```
```ruby
# to access data specify path:
component("ui/forms/select").new(
  ...,
  src: "https://whatcms.org/API/List"
  "data-option-label-field": "label",
  "data-json-path": "result.list"
)
```
### Loading template
A nice little "Loading" template placeholder is there to be displayed while the options are being loaded.

https://github.com/user-attachments/assets/1feaa1fb-49ff-4373-87a6-2c1acca58e01



## Checklist

Check out our [PR guidelines](https://github.com/solidusio/.github/blob/master/CONTRIBUTING.md#pull-request-guidelines) for more details.

The following are mandatory for all PRs:

- [x] [I agree that my PR will be published under the same license as Solidus](https://github.com/solidusio/solidus/blob/main/LICENSE.md).
- [x] I have written a thorough PR description.
- [x] I have kept my commits small and atomic.
- [x] [I have used clear, explanatory commit messages](https://github.com/solidusio/.github/blob/main/CONTRIBUTING.md#writing-good-commit-messages).

tvdeyen commented:
> @chaimann nice! Which select(s) do we need this for?

This is very useful for Product-, Variant- and Taxon-Selects.

tvdeyen commented on admin/app/components/solidus_admin/ui/forms/select/component.rb:
```diff
@@ -25,22 +25,40 @@ class SolidusAdmin::UI::Forms::Select::Component < SolidusAdmin::BaseComponent
   }.freeze
 
-  def initialize(label:, name:, choices:, size: :m, hint: nil, tip: nil, error: nil, **attributes)
+  def initialize(label:, name:, choices: nil, size: :m, hint: nil, tip: nil, error: nil, **attributes)
```
I would prefer to have `src` as named argument instead of "hiding" it in the attributes.

chaimann replied:
makes sense! 😄

tvdeyen reviewed (commented):
Amazing. Thanks. Can we make the `src` argument a bit more approachable? It's too important to "hide" it in a Hash :)

tvdeyen commented on admin/app/javascript/solidus_admin/web_components/solidus_select.js:
```diff
@@ -23,46 +56,58 @@ class SolidusSelect extends HTMLSelectElement {
-        break;
-    }
+  async fetchOptions() {
+    const response = await fetch(this.getAttribute("data-src"));
```
One great thing about custom elements is that you can (and should) add your own attributes without the need to use `data-`

IE:

```html
<solidus-select src="http://..."
```

Just realised that I missed that from the previous PR review :(

chaimann replied:
That's what I was initially planning to do! But at some point I realised that my component is not an autonomous custom one (e.g `<solidus-select>`), but an extended HTML select, and I thought that `data-` attributes would be a more natural use in this case. MDN docs also never mentioned anything about custom attributes, other than `observedAttributes`, but that is obviously for those attributes that can be changed

tvdeyen replied:
Ah. Yes. Of course. Totally ignored that fact. Then data- attributes are absolutely perfect.

tvdeyen reviewed (commented):
can we please not use `data-` attributes on our custom element? Sorry, missed that during previous PR review

chaimann commented:
> @chaimann nice! Which select(s) do we need this for?

@kennyadsl also initially this was a spin-off from PR for zones add/edit where when a zone is state based there is a select with all the states. TomSelect handles dynamically loaded options pretty good, without any noticeable impact in the browsers; there is also a setting `maxOptions` which limits the number of options shown in the dropdown, further improving the performance. I set it to 500 (which I think might even be reduced to 100, but there does not seem to be any difference in performance), and for large option collections I believe the user will still be relying on search filtering rather than manually scrolling to find desired option.

fthobe commented:
@chaimann it's more complicated than that. You can expect on Zip based zones up to thousands of values. 

The issue is outlined in #6187, 
you can expect thousands of values and no effective way to manage them other than fully copy and pasting or csv import (which is not a default option on Solidus). What you did is a great step forward in allowing effective management of zones.

chaimann commented:
@fthobe amazing! I'm super glad that it's going to be useful :)

tvdeyen commented on admin/app/javascript/solidus_admin/web_components/solidus_select.js:
```diff
@@ -105,9 +105,18 @@ class SolidusSelect extends HTMLSelectElement {
   async fetchOptions() {
-    const response = await fetch(this.getAttribute("data-src"));
-    return await response.json();
+    const [url, dataPath] = this.getAttribute("data-src").split("::");
```
No. But it seems a bit weird. I think another attribute is nothing bad.

chaimann replied:
sure. I had another thought and guess being more explicit about this feature might be a better choice

tvdeyen commented on admin/app/javascript/solidus_admin/web_components/solidus_select.js:
```diff
@@ -5,15 +5,48 @@ class SolidusSelect extends HTMLSelectElement {
       optionClass: "option",
       wrapperClass: "wrapper",
-      maxOptions: null,
+      maxOptions: 500,
```
Is there a way to utilize the pagination feature from our API, like we do in the [productPicker](https://github.com/solidusio/solidus/blob/63d01f1c33d074e51ed71e8a35420ebe8f00a2a5/backend/app/assets/javascripts/spree/backend/product_picker.js#L54) and [variantAutoComplete](https://github.com/solidusio/solidus/blob/63d01f1c33d074e51ed71e8a35420ebe8f00a2a5/backend/app/assets/javascripts/spree/backend/variant_autocomplete.js#L48)?

chaimann replied:
@tvdeyen The results are actually preloaded in full, and this `maxOptions` does not control how many results we pull, but how many results we _display_ on dropdown open. It's like a small "hack" that allows to reduce the lag on TomSelect side (when displaying more than 1k options the search and select/unselect starts to have a visible delay). As I mentioned previously, having it in place helps to build a visual of a "full" dropdown, while expecting the user to use search instead of selecting options by manually scroll-finding them.

We also should be able to make pagination work with TomSelect, there is a plugin for that which I haven't tested yet but I believe it should work. I agree that it would be better in the long term to have paginated results instead of full preload, as the larger a dataset is the longer the request would be processed. The only downside I can see is that the search function would now need to hit the server as well, but I think I can configure it to search the "cached" fetched data first.

tvdeyen reviewed (commented):
I really like that! I also worry about the 500 results we pull. Can we utilize the pagination for that? Does TomSelect support it, like select2 does?

chaimann commented:
https://github.com/user-attachments/assets/7ac96387-f62c-437c-9b19-1798d5da27f4

Added infinite scroll for remote select, also patched the bug with dropdown scroll, and redesigned single select search to show search term when typing in the field, restoring selected option if user does not select anything after search.
Also cleaned up "solidus_select" by separating different pieces of logic into separate plugins, courtesy of TomSelect plugins system :)

tvdeyen commented:
Amazing. Will take a closer look tonight, as I am currently not at a computer

chaimann commented:
Thanks Thomas! I forgot to add that there is a visual bug with TomSelect when loading more options it will mark any option you hover on as "active":

https://github.com/user-attachments/assets/4b7fa04b-510e-40fd-bb73-4ca530ac319d

It does not affect functionality though, for example on the video I clicked "enter" to select a different option and it selected it correctly, so it's just a problem with TomSelect not removing "active" class on those options while loading. I decided not to spend more time trying to patch it (maybe I'll return to it later, just wanted to switch to different tasks already 😄)

tvdeyen commented on admin/app/javascript/solidus_admin/tom-select/plugins/remote_with_pagination.js:
```diff
@@ -0,0 +1,67 @@
+      return "<div class='loading'>Loading</div>";
+    },
+    loading_more: function() {
+      return "<div class='loading-more'>Loading more results</div>";
```
The idea to be able to override the messages HTML is neat, but I am very certain that this is rarely used (if ever at all). 

How are you imaging localization of this string? Can you give an example how - for instance `solidus_i18n` - would add a translation here? I am guessing it is expected that gems and apps override the select component then? Seems a bit overkill to me for a simple string we want to pass from Ruby to HTML.

I think a more approachable solution could be to introduce two new options passed to the Select View Component (ie. `loading_message` and `load_more_message`) with defaults passing the strings to `I18n` with a proper scope.

WDYT?

tvdeyen commented on admin/app/javascript/solidus_admin/tom-select/plugins/stash_on_search.js:
```diff
@@ -0,0 +1,70 @@
+// Instead, hide the selected option when typing and bring it back when leaving the field.
+class Stash {
+  constructor() {
+    this.content = {};
```
Nit: You could [make use of a private field here](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Classes/Private_properties).

tvdeyen commented on admin/app/javascript/solidus_admin/tom-select/plugins/stash_on_search.js:
```diff
@@ -0,0 +1,70 @@
+    this.content = {};
+  }
+
+  isEmpty() {
```
Nit: A getter would be useful here.

tvdeyen reviewed (commented):
Wow. Amazing work 👏🏻 

I have some minor nits about embracing modern JS features, but the main concern I have is with localization of the loading messages. I think we can simplify this to just passing strings to I18n and use that in the custom element.

BTW. This has been more work than I expected by using a library like tom-select. But I think it is worth it and a major improvement. 

Thanks for investing all this time.

PPS: We should think about setting up JS testing soon, so we can test our code in the future.

tvdeyen commented on admin/app/javascript/solidus_admin/web_components/solidus_select.js:
```diff
@@ -64,6 +36,41 @@ class SolidusSelect extends HTMLSelectElement {
     }
   }
+
+  getTomSelectSettings() {
```
Woops. The comment I made oin this file went missing somehow (thanks github. react is not an improvement btw...)

We could use a JS getter here.

```suggestion
  get tomSelectSettings() {
```

tvdeyen replied:
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/get

chaimann replied:
thanks I completely forgot about js getters
